src/codegen/polkadot.rs

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Contains `codegen` helpers for the Polkadot target.
+
+use solang_parser::pt::Loc;
+
+use crate::{
+    codegen::{
+        cfg::{ControlFlowGraph, Instr},
+        revert::log_runtime_error,
+        vartable::Vartable,
+        Expression, Options,
+    },
+    sema::ast::{Namespace, Type},
+};
+
+/// Helper to handle error cases from external function and constructor calls.
+pub(crate) struct RetCodeCheck {
+    pub success: usize,
+    pub revert: usize,
+    pub error_no_data: usize,
+    msg: &'static str,
+    loc: Loc,
+}
+
+#[derive(Default)]
+pub(crate) struct RetCodeCheckBuilder<Success = ()> {
```

Comment:
Why the generic parameter when it is always going to be a `usize`?

Reply:
This is a typestate builder

Reply:
Well, the typestate enforces 1) that the `success` var was set, as the impl of <success = ()> doesn't implement `insert` to conclude the builder (only `<Success = usize> does implement it) and 2) allow exact behavior regarding to at what state which variable is allowed to be set, all at compile time. I implemented it this way originally because I thought it was necessary while it's actually not. But I'll leave it in place now, I mean it's all compile time and makes for an easily extendable interface.